src/script.rs

```diff
@@ -0,0 +1,642 @@
+    fn try_from(value: ScriptNum) -> Result<Self, Self::Error> {
+        value.getint().try_into()
+    }
+}
+
+impl TryFrom<ScriptNum> for u8 {
+    type Error = TryFromIntError;
+    fn try_from(value: ScriptNum) -> Result<Self, Self::Error> {
+        value.getint().try_into()
+    }
+}
+
+impl Add for ScriptNum {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        let rhs = other.0;
+        assert!(
+            rhs == 0
+                || (rhs > 0 && self.0 <= i64::MAX - rhs)
+                || (rhs < 0 && self.0 >= i64::MIN - rhs)
+        );
+        Self(self.0 + rhs)
+    }
+}
+
+impl Sub for ScriptNum {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        let rhs = other.0;
+        assert!(
+            rhs == 0
+                || (rhs > 0 && self.0 >= i64::MIN + rhs)
+                || (rhs < 0 && self.0 <= i64::MAX + rhs)
+        );
+        Self(self.0 - rhs)
+    }
+}
+
+impl Neg for ScriptNum {
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        assert!(self.0 != i64::MIN);
+        Self(-self.0)
+    }
+}
+
+/** Serialized script, used inside transaction inputs and outputs */
+#[derive(Clone, Debug)]
+pub struct Script<'a>(pub &'a [u8]);
+
+impl Script<'_> {
+    pub fn get_op(script: &mut &[u8]) -> Result<Opcode, ScriptError> {
+        Self::get_op2(script, &mut vec![])
+    }
+
+    pub fn get_op2(script: &mut &[u8], buffer: &mut Vec<u8>) -> Result<Opcode, ScriptError> {
+        if script.is_empty() {
+            return Err(ScriptError::ReadError {
+                expected_bytes: 1,
+                available_bytes: 0,
+            });
+        }
+
+        // Empty the provided buffer, if any
+        buffer.truncate(0);
+
+        let leading_byte = Opcode::from(script[0]);
+        *script = &script[1..];
+
+        Ok(match leading_byte {
+            Opcode::PushValue(pv) => match pv {
+                OP_PUSHDATA1 | OP_PUSHDATA2 | OP_PUSHDATA4 => {
+                    let read_le = |script: &mut &[u8], needed_bytes: usize| {
+                        if script.len() < needed_bytes {
+                            Err(ScriptError::ReadError {
+                                expected_bytes: needed_bytes,
+                                available_bytes: script.len(),
+                            })
+                        } else {
+                            let mut size = 0;
+                            for i in (0..needed_bytes).rev() {
+                                size <<= 8;
+                                size |= usize::from(script[i]);
+                            }
+                            *script = &script[needed_bytes..];
+                            Ok(size)
+                        }
+                    };
+
+                    let size = match pv {
+                        OP_PUSHDATA1 => read_le(script, 1),
+                        OP_PUSHDATA2 => read_le(script, 2),
+                        OP_PUSHDATA4 => read_le(script, 4),
```

Comment:
```suggestion
                        OP_PUSHDATA1 => usize::from(script[0]),
                        OP_PUSHDATA2 => usize::from(u16::from_le_bytes(script.try_into().unwrap())),
                        OP_PUSHDATA4 => usize::from(u32::from_le_bytes(script.try_into().unwrap())),
```
and delete `read_le`. (A further refactoring could be to put the code below in a closure so that you don't need the nested match.)

Reply:
With the other changes that I backported, this is now handled with a nice `split_at_checked`.